contracts/prize-pool/barnbridge/BarnPrizePool.sol

```diff
@@ -0,0 +1,101 @@
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+
+import "../../external/barnbridge/BarnInterface.sol";
+import "../PrizePool.sol";
+
+/// @title Prize Pool for Barn Bridge $BOND token
+contract BarnPrizePool is PrizePool {
+  using SafeMathUpgradeable for uint256;
+
+  event BarnPrizePoolInitialized(address indexed barn);
+
+  /// @notice Interface for the barn
+  BarnInterface public barn;
+
+  /// @notice Initializes the Prize Pool and Yield Service with the required contract connections
+  /// @param _controlledTokens Array of addresses for the Ticket and Sponsorship Tokens controlled by the Prize Pool
+  /// @param _maxExitFeeMantissa The maximum exit fee size, relative to the withdrawal amount
+  /// @param _maxTimelockDuration The maximum length of time the withdraw timelock could be
+  /// @param _barn Address of the barn
+  function initialize (
+    RegistryInterface _reserveRegistry,
+    ControlledTokenInterface[] memory _controlledTokens,
+    uint256 _maxExitFeeMantissa,
+    uint256 _maxTimelockDuration,
+    BarnInterface _barn
+  )
+    public
+    initializer
+  {
+    PrizePool.initialize(
+      _reserveRegistry,
+      _controlledTokens,
+      _maxExitFeeMantissa,
+      _maxTimelockDuration
+    );
+    barn = _barn;
+
+    emit BarnPrizePoolInitialized(address(barn));
+  }
+
+  /// @dev Gets the balance of the underlying assets held by the Yield Service
+  /// @return The underlying balance of asset tokens
+  function _balance() internal override returns (uint256) {
+    uint256 balance = barn.balanceOf(address(this));
+    return balance;
+  }
+
+  /// @dev Allows a user to supply asset tokens in exchange for yield-bearing tokens
+  /// to be held in escrow by the Yield Service
+  function _supply(uint256) internal override {
+    IERC20Upgradeable assetToken = _token();
+    uint256 total = assetToken.balanceOf(address(this));
+    assetToken.approve(address(barn), total);
+    barn.deposit(total);
+  }
+
+  /// @dev Allows a user to supply asset tokens in exchange for yield-bearing tokens
+  /// to be held in escrow by the Yield Service
+  function _supplySpecific(uint256 amount) internal {
+    _token().approve(address(barn), amount);
+    barn.deposit(amount);
+  }
+
+  /// @dev The external token cannot be yDai or Dai
+  /// @param _externalToken The address of the token to check
+  /// @return True if the token may be awarded, false otherwise
+  function _canAwardExternal(address _externalToken) internal override view returns (bool) {
+    return _externalToken != address(barn) && _externalToken != address(barn.token());
+  }
+
+  /// @dev Allows a user to redeem yield-bearing tokens in exchange for the underlying
+  /// asset tokens held in escrow by the Yield Service
+  /// @param amount The amount of underlying tokens to be redeemed
+  /// @return The actual amount of tokens transferred
+  function _redeem(uint256 amount) internal override returns (uint256) {
```

Comment:
I think this logic might be a bit too simple and is not accounting for early exits, but it all depends on the staking protocol behavior.
Lets say the lottery has 7 more days to go and someone stakes 1 BARN. On day 6 (1 day before the lottery end) this 1 BARN has accrued 0.1 BARN. It is very important that the user if the user decides to exit, he is given back maximum of 1 BARN and the 0.1 BARN accrued is returned to the staking protocol and eventually gotten by the winner. Can you check and see if this is the case?

Reply:
Yes, I agree. Basically in the protocol - there are two possibilities - deposit $BOND without locking it (nothing is accrued in this case) and deposit $BOND and lock it (maximum duration 1 year, where accrued amount is linear to the time staked, with maximum multiplier x2). However, when $BOND is locked, it cannot be withdrawn until the time has passed, so basically from what I see, the user wouldn't be able to make an early exit.